### ISSUE TITLE
ansible_local var no longer available.

### ISSUE THREAD

**Issue by sudochop** [ansible/ansible#14641]
##### ISSUE TYPE
Bug Report

##### COMPONENT NAME
core

##### ANSIBLE VERSION
```
ansible 1.9.4
configured module search path = None
```

##### OS / ENVIRONMENT
OSX to CentOS

##### SUMMARY

Since upgrading, my machine specific facts cannot be found. I am setting facts in `/etc/ansible/facts.d`, and since upgrading, I now just receive this error when trying to access them in my template:
```
{'msg': "AnsibleUndefinedVariable: One or more undefined variables: 'dict object' has no attribute 'ansible_local'", 'failed': True}
```
The part of my jinja template in question:
```
{% for host in groups['ec2'] %}
{{ hostvars[host]['ansible_default_ipv4']['address'] }} {{ hostvars[host]['ansible_hostname'] }} {{ hostvars[host]['ansible_local']['app']['hostname'] }} 
{% endfor %}
```
I'm not entirely sure which version I was running before upgrading to latest - but this immediately stopped working after upgrading.

**Comment by bsdlp** [comment]
do we know which version started breaking?

**Comment by sudochop** [comment]
Unfortunately not. I couldn't have been far behind.

**Comment by lunetics** [comment]
I'm using 2.2 and this one is happening on one of our servers, so i guess it's some python issue:

fatal: [backup.xxx.com]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'ansible_local' is undefined"}

**Comment by macropin** [comment]
I think this issue relates to facts collection. If you have not collected facts for all the hosts then this issue may come up.

Try running `ansible all -m setup` to collect facts for all hosts.

**Comment by ryansb** [comment]
Can you share the facts you have in `facts.d` that are getting loaded, and the output when running your test with `-vvvv` enabled?

**Comment by ryansb** [comment]
needs_info

**Comment by ansibot** [comment]
@sudochop ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@sudochop ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by towolf** [comment]
I just encountered the problem, where `ansible_local` is lost between the top-level playbook where gathering happens (debug: var=ansible_local looks good in pre_tasks) and when I want to access it in a sub-sub-role (debug: var=ansible_local in the role shows empty `{}`).

When I re-gather facts in the sub-role (with setup:) immediately before, then it works again.

This started after 2.2.1.

@ryansb

**Comment by ansibot** [comment]
@sudochop Greetings! Thanks for taking the time to open this issue. In order for the community to handle your issue effectively, we need a bit more information. 

Here are the items we could not find in your description:
- component name

Please set the description of this issue with this template:
https://raw.githubusercontent.com/ansible/ansible/devel/.github/ISSUE_TEMPLATE.md

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jamiejamison** [comment]
I'm having the same problem in ansible version 2.2.2.0-1 on Ubuntu Trusty.

**Comment by bcoca** [comment]
I cannot reproduce locally, can anyone provide a simple reproducer in a current version of Ansible?

**Comment by towolf** [comment]
@bcoca, please see this minimal repro repo: https://github.com/towolf/ansible_local_empty

**Comment by towolf** [comment]
Ok, I just realized something weird.

When I comment out the `debug:` in the top level playbook, then the second `debug:` in role1 works.

And even worse, when I put two identical `debug:` tasks right after each other, the second one is empty.

```
  pre_tasks:                                                                                                                                                                                      
    - debug:                                                                                                                                                                                      
        var: ansible_local                                                                                                                                                                        
    - debug:                                                                                                                                                                                      
        var: ansible_local 
```

outputs
```
TASK [debug] *************************************************************************************************************************************************************************************
ok: [localhost] => {
    "ansible_local": {
        "my": {
            "greeting": {
                "message": "\"Hi, Brian\""
            }
        }
    }, 
    "changed": false
}

TASK [debug] *************************************************************************************************************************************************************************************
ok: [localhost] => {
    "ansible_local": {}, 
    "changed": false
}
```

**Comment by towolf** [comment]
Even this does not work:

```
  tasks:                                                                                                                                                                                          
    - debug:                                                                                                                                                                                      
        var: inventory_hostname                                                                                                                                                                   
    - debug:                                                                                                                                                                                      
        var: ansible_local  
```

```
PLAY [Why is ansible_local lost?] ********

TASK [Gathering Facts] ***********************************
ok: [localhost]

TASK [debug] ******************
ok: [localhost] => {
    "changed": false, 
    "inventory_hostname": "localhost"
}

TASK [debug] *************
ok: [localhost] => {
    "ansible_local": {}, 
    "changed": false
}

PLAY RECAP *********************
localhost      
```

**Comment by towolf** [comment]
@bcoca, seems to be this: 97cb2016d86523ad6efbee2c2975ca287075a2a4

But now I'm puzzled because my problem was with 2.2.2.

**Comment by bcoca** [comment]
so back to needing info, anyone that can reproduce this on a current version?

**Comment by apenney** [comment]
I can, @bcoca:

```
» ansible --version
ansible 2.2.3.0 (stable-2.2 7ff9fa52cf) last updated 2017/04/10 20:37:40 (GMT -400)
  lib/ansible/modules/core: (detached HEAD cfe3c844f4) last updated 2017/04/11 10:00:41 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 8b78d56a08) last updated 2017/04/11 10:00:41 (GMT -400)
```

```
    amazon-ebs: TASK [silpion.util : Assert EPEL configuration] ********************************
    amazon-ebs: fatal: [default]: FAILED! => {"failed": true, "msg": "The conditional check 'ansible_local.util.epel.enable is defined' failed. The error was: error while evaluating conditional (ansible_local.util.epel.enable is defined): 'ansible_local' is undefined\n\nThe error appears to have been in '/Users/apenney/maxwell/ansible/roles/silpion.util/tasks/main.yml': line 56, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n# This protects the role in various edge cases to function properly.\n- name: Assert EPEL configuration\n  ^ here\n"}
```

**Comment by dnmvisser** [comment]
I think I just hit this as well while running a playbook that includes the `silpion.util` role, I haven't touched the playbook for a while but now on 2.2.2 it gives me:
```
TASK [silpion.util : Assert EPEL configuration] ********************************
fatal: [eventr-prod]: FAILED! => {"failed": true, "msg": "The conditional check 'ansible_local.util.epel.enable is defined' failed. The error was: error while evaluating conditional (ansible_local.util.epel.enable is defined): 'ansible_local' is undefined\n\nThe error appears to have been in '/Users/dick.visser/ansible/roles/silpion.util/tasks/main.yml': line 56, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n# This protects the role in various edge cases to function properly.\n- name: Assert EPEL configuration\n  ^ here\n"}
```
Just switched back to 2.2.1 and then problem solved...

**Comment by hloeffler** [comment]
@bcoca

**Comment by bcoca** [comment]
can you show full play and how ansible local files are defined?

**Comment by gordogre** [comment]
I also ran into this when running silpion.util, and I feel certain that the problem is that the silpion.util tasks frequently check for the existence of a particular fact, for example:

`when: ansible_local.util.epel.enable is defined`

However, as of ansible 2.2.2, performing this check fails if the parent properties themselves are undefined.

I was also able to revert to 2.2.1 and get the silpion.util working. 

I had also started modifying the silpion.util to change those conditional checks to first check for the existence of the parent, but there are quite a few of them, and that becomes messy, and more importantly it feels like an un-intended change to ansible's conditional checking.

**Comment by bcoca** [comment]
Someone in irc just mentioned that it only happens if 'networking' (fixed when !networking is used in fact subsets) is involved ... do all of you have a network interface named 'local'?

**Comment by dericcrago** [comment]
I believe I'm running into this issue (or something very similar) as well. Here's an arbitrary example I came up with to show how it's changed over the last couple of versions. Is this helpful @bcoca?

```yaml
# local_facts.yml
---
- hosts: all
  connection: local
  tasks:
    - debug:
        var: ansible_local

    - debug:
        msg: RAN because ansible_local is undefined!
      when: ansible_local is undefined

    - debug:
        var: ansible_local.foo

    - debug:
        msg: RAN because ansible_local.foo is undefined!
      when: ansible_local.foo is undefined

    - debug:
        var: ansible_local.foo.bar

    - debug:
        msg: RAN because ansible_local.foo.bar is undefined!
      when: ansible_local.foo.bar is undefined

```

```bash
# ansible --version
ansible 2.2.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

# ansible-playbook -i localhost, local_facts.yml

PLAY [all] *********************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] => {
    "ansible_local": "VARIABLE IS NOT DEFINED!"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "RAN because ansible_local is undefined!"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "ansible_local.foo": "VARIABLE IS NOT DEFINED!"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "RAN because ansible_local.foo is undefined!"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "ansible_local.foo.bar": "VARIABLE IS NOT DEFINED!"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "RAN because ansible_local.foo.bar is undefined!"
}

PLAY RECAP *********************************************************************
localhost                  : ok=7    changed=0    unreachable=0    failed=0   

```

```bash
# ansible --version
ansible 2.2.2.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

# ansible-playbook -i localhost, local_facts.yml

PLAY [all] *********************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] => {
    "ansible_local": "VARIABLE IS NOT DEFINED!"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "RAN because ansible_local is undefined!"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "ansible_local.foo": "VARIABLE IS NOT DEFINED!"
}

TASK [debug] *******************************************************************
fatal: [localhost]: FAILED! => {"failed": true, "msg": "The conditional check 'ansible_local.foo is undefined' failed. The error was: error while evaluating conditional (ansible_local.foo is undefined): 'ansible_local' is undefined\n\nThe error appears to have been in '/local_facts.yml': line 16, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n    - debug:\n      ^ here\n"}
  to retry, use: --limit @/local_facts.retry

PLAY RECAP *********************************************************************
localhost                  : ok=4    changed=0    unreachable=0    failed=1   

```

```bash
# ansible --version
ansible 2.4.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /venv/ansible/lib/python2.7/site-packages/ansible
  executable location = /venv/ansible/bin/ansible
  python version = 2.7.13 (default, Jun  1 2017, 17:33:32) [GCC 4.4.7 20120313 (Red Hat 4.4.7-18)]

# ansible-playbook -i localhost, local_facts.yml

PLAY [all] *********************************************************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] *********************************************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [debug] *******************************************************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "ansible_local": {}
}

TASK [debug] *******************************************************************************************************************************************************************************************************************************************************************
skipping: [localhost]

TASK [debug] *******************************************************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "ansible_local.foo": "VARIABLE IS NOT DEFINED!"
}

TASK [debug] *******************************************************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "RAN because ansible_local.foo is undefined!"
}

TASK [debug] *******************************************************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "ansible_local.foo.bar": "VARIABLE IS NOT DEFINED!"
}

TASK [debug] *******************************************************************************************************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"msg": "The conditional check 'ansible_local.foo.bar is undefined' failed. The error was: error while evaluating conditional (ansible_local.foo.bar is undefined): 'dict object' has no attribute 'foo'\n\nThe error appears to have been in '/local_facts.yml': line 23, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n    - debug:\n      ^ here\n"}
  to retry, use: --limit @/local_facts.retry

PLAY RECAP *********************************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=5    changed=0    unreachable=0    failed=1   

```

**Comment by bcoca** [comment]
in 2.4 its always defined, even if empty, that is new ... but do you have anything in /etc/facts.d that would populate it?

**Comment by dericcrago** [comment]
Hi @bcoca, I think I may be conflating issues. The problem I'm experiencing isn't that they're undefined, I'm expecting that, it's that the `when`s behave differently between versions. The task with `when: ansible_local.foo.bar is undefined` runs on ansible 2.2.1.0, but not on later versions. I think I could get around this by building up the when, `when: ansible_local is undefined or ansible_local.foo is undefined or ansible_local.foo.bar is undefined` see below for an example. Should this be a separate issue?

```bash
# ansible --version
ansible 2.2.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

# ansible-playbook -i localhost, local_facts.yml

PLAY [all] *********************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] => {
    "ansible_local": "VARIABLE IS NOT DEFINED!"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "ansible_local.foo": "VARIABLE IS NOT DEFINED!"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "ansible_local.foo.bar": "VARIABLE IS NOT DEFINED!"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "RAN because ansible_local is undefined!"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "RAN because ansible_local is undefined or ansible_local.foo is undefined!"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "RAN because ansible_local is undefined or ansible_local.foo is undefined or ansible_local.foo.bar is undefined!"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "RAN because ansible_local.foo is undefined!"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "RAN because ansible_local.foo.bar is undefined!"
}

PLAY RECAP *********************************************************************
localhost                  : ok=9    changed=0    unreachable=0    failed=0   
```

```bash
# ansible --version
ansible 2.2.2.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

# ansible-playbook -i localhost, local_facts.yml

PLAY [all] *********************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [debug] *******************************************************************
ok: [localhost] => {
    "ansible_local": "VARIABLE IS NOT DEFINED!"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "ansible_local.foo": "VARIABLE IS NOT DEFINED!"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "ansible_local.foo.bar": "VARIABLE IS NOT DEFINED!"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "RAN because ansible_local is undefined!"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "RAN because ansible_local is undefined or ansible_local.foo is undefined!"
}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "RAN because ansible_local is undefined or ansible_local.foo is undefined or ansible_local.foo.bar is undefined!"
}

TASK [debug] *******************************************************************
fatal: [localhost]: FAILED! => {"failed": true, "msg": "The conditional check 'ansible_local.foo is undefined' failed. The error was: error while evaluating conditional (ansible_local.foo is undefined): 'ansible_local' is undefined\n\nThe error appears to have been in '/local_facts.yml': line 27, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n    - debug:\n      ^ here\n"}
    to retry, use: --limit @/local_facts.retry

PLAY RECAP *********************************************************************
localhost                  : ok=7    changed=0    unreachable=0    failed=1   
```

```bash
# ansible --version
ansible 2.4.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /venv/ansible/lib/python2.7/site-packages/ansible
  executable location = /venv/ansible/bin/ansible
  python version = 2.7.13 (default, Jun  1 2017, 17:33:32) [GCC 4.4.7 20120313 (Red Hat 4.4.7-18)]

# ansible-playbook -i localhost, local_facts.yml

PLAY [all] *************************************************************************************************************************************************************************************************

TASK [Gathering Facts] *************************************************************************************************************************************************************************************
ok: [localhost]

TASK [debug] ***********************************************************************************************************************************************************************************************
ok: [localhost] => {
    "ansible_local": {}
}

TASK [debug] ***********************************************************************************************************************************************************************************************
ok: [localhost] => {
    "ansible_local.foo": "VARIABLE IS NOT DEFINED!"
}

TASK [debug] ***********************************************************************************************************************************************************************************************
ok: [localhost] => {
    "ansible_local.foo.bar": "VARIABLE IS NOT DEFINED!"
}

TASK [debug] ***********************************************************************************************************************************************************************************************
skipping: [localhost]

TASK [debug] ***********************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "RAN because ansible_local is undefined or ansible_local.foo is undefined!"
}

TASK [debug] ***********************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "RAN because ansible_local is undefined or ansible_local.foo is undefined or ansible_local.foo.bar is undefined!"
}

TASK [debug] ***********************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "RAN because ansible_local.foo is undefined!"
}

TASK [debug] ***********************************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"msg": "The conditional check 'ansible_local.foo.bar is undefined' failed. The error was: error while evaluating conditional (ansible_local.foo.bar is undefined): 'dict object' has no attribute 'foo'\n\nThe error appears to have been in '/local_facts.yml': line 31, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n    - debug:\n      ^ here\n"}
    to retry, use: --limit @/local_facts.retry

PLAY RECAP *************************************************************************************************************************************************************************************************
localhost                  : ok=7    changed=0    unreachable=0    failed=1   
```

**Comment by ansibot** [comment]
@sudochop This issue is waiting for your response. Please respond or the issue will be closed.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by sudochop** [comment]
Bad bot.

**Comment by curtisforrester** [comment]
I also am seeing this with Ansible 2.5 - though slightly different failure. It's worth noting that I only see this when gathering facts for the local machine, but plays for a remote server behave normally as expected.  I even tried to reference the local machine by name/SSH and still no facts.

Using the playbook that @dericcrago posted above, the output is:

```
PLAY [control] *****************************************************************************************************************************************************************************************

TASK [Gathering Facts] *********************************************************************************************************************************************************************************
ok: [control]

TASK [debug] *******************************************************************************************************************************************************************************************
ok: [control] => {
    "ansible_local": {}
}

TASK [debug] *******************************************************************************************************************************************************************************************
skipping: [control]

TASK [debug] *******************************************************************************************************************************************************************************************
ok: [control] => {
    "ansible_local.foo": "VARIABLE IS NOT DEFINED!"
}

TASK [debug] *******************************************************************************************************************************************************************************************
ok: [control] => {
    "msg": "RAN because ansible_local.foo is undefined!"
}

TASK [debug] *******************************************************************************************************************************************************************************************
ok: [control] => {
    "ansible_local.foo.bar": "VARIABLE IS NOT DEFINED!"
}

TASK [debug] *******************************************************************************************************************************************************************************************
fatal: [control]: FAILED! => {"msg": "The conditional check 'ansible_local.foo.bar is undefined' failed. The error was: error while evaluating conditional (ansible_local.foo.bar is undefined): 'dict object' has no attribute 'foo'\n\n
...snip...

PLAY RECAP *********************************************************************************************************************************************************************************************
control                    : ok=5    changed=0    unreachable=0    failed=1  
```

**Comment by juniorx32g** [comment]
I'm using 2.2.1 and noticed that with some of my hosts, this issue only happened when I didn't pass the `-b` option.

_updating_: So, in my case, I found out the issue was with the directory `/etc/ansible` missing a `x` permission for all users. Fixed it with a `chmod 755 /etc/ansible` in the target host.

- my test playbook:
```
---
- hosts: "{{ host }}"
  gather_facts: yes
  tasks:
    - debug: var=ansible_local
```
- without `-b`
```
> $ ansible-playbook test-playbook.yml -i inventories/ -e host=some_host 

PLAY [some_host] *********************************************************************

TASK [setup] *******************************************************************
ok: [some_host]

TASK [debug] *******************************************************************
ok: [some_host] => {
    "ansible_local": "VARIABLE IS NOT DEFINED!"
}

PLAY RECAP *********************************************************************
some_host                        : ok=2    changed=0    unreachable=0    failed=0   
```

- and with `-b`
```
> $ ansible-playbook test-playbook.yml -i inventories/ -e host=some_host -b 

PLAY [some_host] *********************************************************************

TASK [setup] *******************************************************************
ok: [some_host]

TASK [debug] *******************************************************************
ok: [some_host] => {
    "ansible_local": {
(....)
}

PLAY RECAP *********************************************************************
some_host                        : ok=2    changed=0    unreachable=0    failed=0   
```

**Comment by bcoca** [comment]
@dericcrago  yes, that seems like a different issue, I have added a patch that ensures `ansible_local` is always defined, even if empty, so it might not be as big of an issue anymore.

@juniorx32g so that looks like it is a permissions issue, the 'fact gathering' must be able to read/execute the things you put in the 'local' dir on the remote.

I've been unable to reproduce a problem with the code from all the examples given,  most seem to be explainable due to missing permissions or incorrect file or some other matter tangentially related. So for now I'm going to close this issue.

If anyone still hits this kind of problem and is sure it is not a permissions/format issue, feel free to open new issues with a redproducer.

**Comment by kumareshan11** [comment]
Am encountering same issue with below Ansible version :

Ansible version : 
```
ansible 2.8.2
python version = 3.7.0 (default, May 29 2019, 23:43:42) [GCC 5.4.0 20160609]
```

playbook : 
---
```
- hosts: all 
  become: yes 
  tasks:
    - name: install custom facts
      include_role:
        name: facts

- hosts: app-cluster
  become: yes 
  gather_facts: yes 
  tasks:    
    - name: install app
      include_role:
        name: app
      vars:
        user: "{{ ansible_user | default(ansible_env.SUDO_USER) }}"
```

related app role : 
------------------
```
- debug: var=ansible_local

- name: copy base configuration files
  template:
    src: "{{ item }}"
    dest: "/etc/app/{{ item }}"
  with_items:
    - app_base.conf
    - app_top.conf
  notify:
    - restart service
  tags:
    - config

```


facts debug output & error : 
=======
```
debug...
  node-01 ok: {
    "ansible_local": {
        "gpus": {
            "count": 0
        },
        "memory": {
            "total_mb": 182489
        },
    "changed": false
}
..
...
....
  node-01 failed: {
    "ansible_loop_var": "item",
    "changed": false,
    "item": "app.conf",
    "msg": "AnsibleUndefinedVariable: 'ansible.vars.hostvars.HostVarsVars object' has no attribute 'ansible_local'"
}
```
Template : 
```
{% set memory =  hostvars[node_name]["ansible_local"]["memory"] -%}

```
It fails to get ansible_local vars 


When i run the play on all ansible nodes, i see no issue with getting "ansible_local" facts but when i target single node. It countering this issue and am able to see the " ansible_local" facts output with debug option in same playbook before this task exeute.

**Comment by kumareshan11** [comment]
> @dericcrago yes, that seems like a different issue, I have added a patch that ensures `ansible_local` is always defined, even if empty, so it might not be as big of an issue anymore.
> 
> @juniorx32g so that looks like it is a permissions issue, the 'fact gathering' must be able to read/execute the things you put in the 'local' dir on the remote.
> 
> I've been unable to reproduce a problem with the code from all the examples given, most seem to be explainable due to missing permissions or incorrect file or some other matter tangentially related. So for now I'm going to close this issue.
> 
> If anyone still hits this kind of problem and is sure it is not a permissions/format issue, feel free to open new issues with a redproducer.

Am encountering same issue with below Ansible version :
OS : 
NAME="Ubuntu"
VERSION="18.04.2 LTS (Bionic Beaver)"
Ansible version : 
```
ansible 2.8.2
python version = 3.7.0 (default, May 29 2019, 23:43:42) [GCC 5.4.0 20160609]
```

playbook : 
---
```
- hosts: all 
  become: yes 
  tasks:
    - name: install custom facts
      include_role:
        name: facts

- hosts: app-cluster
  become: yes 
  gather_facts: yes 
  tasks:    
    - name: install app
      include_role:
        name: app
      vars:
        user: "{{ ansible_user | default(ansible_env.SUDO_USER) }}"
```

related app role : 
------------------
```
- debug: var=ansible_local

- name: copy base configuration files
  template:
    src: "{{ item }}"
    dest: "/etc/app/{{ item }}"
  with_items:
    - app_base.conf
    - app_top.conf
  notify:
    - restart service
  tags:
    - config

```


facts debug output & error : 
=======
```
debug...
  node-01 ok: {
    "ansible_local": {
        "gpus": {
            "count": 0
        },
        "memory": {
            "total_mb": 182489
        },
    "changed": false
}
..
...
....
  node-01 failed: {
    "ansible_loop_var": "item",
    "changed": false,
    "item": "app.conf",
    "msg": "AnsibleUndefinedVariable: 'ansible.vars.hostvars.HostVarsVars object' has no attribute 'ansible_local'"
}
```
Template : 
```
{% set memory =  hostvars[node_name]["ansible_local"]["memory"] -%}

```
It fails to get ansible_local vars 


When i run the play on all ansible nodes, i see no issue with getting "ansible_local" facts but when i target single node. It countering this issue and am able to see the " ansible_local" facts output with debug option in same playbook before this task executes.
